jax/lax/lax_fft.py

```diff
@@ -16,42 +16,117 @@
 from __future__ import division
 from __future__ import print_function
 
+from functools import partial
+
 import numpy as onp
 
 from jax.abstract_arrays import ShapedArray
+from jax.api import jit, vjp
 from jax.core import Primitive
 from jax.interpreters import xla
-from .. import dtypes
+from . import dtypes, lax
+from ..lib.xla_bridge import xla_client
 from ..interpreters import ad
 from ..interpreters import batching
 
 
-def fft(x, fft_type, fft_lengths=None):
-  if fft_lengths is None:
-    fft_lengths = x.shape
-  elif len(fft_lengths) == 0:
+def _promote_to_complex(arg):
+  dtype = onp.result_type(arg, onp.complex64)
+  # XLA's FFT op only supports C64.
+  if dtype == onp.complex128:
+    dtype = onp.complex64
+  return lax.convert_element_type(arg, dtype)
+
+def _promote_to_real(arg):
+  dtype = onp.result_type(arg, onp.float64)
+  # XLA's FFT op only supports F32.
+  if dtype == onp.float64:
+    dtype = onp.float32
+  return lax.convert_element_type(arg, dtype)
+
+def fft(x, fft_type, fft_lengths):
+  if fft_type == xla_client.FftType.RFFT:
+    if onp.iscomplexobj(x):
+      raise ValueError("only real valued inputs supported for rfft")
+    x = _promote_to_real(x)
+  else:
+    x = _promote_to_complex(x)
+  if len(fft_lengths) == 0:
     # XLA FFT doesn't support 0-rank.
     return x
-  else:
-    fft_lengths = tuple(fft_lengths)
+  fft_lengths = tuple(fft_lengths)
   return fft_p.bind(x, fft_type=fft_type, fft_lengths=fft_lengths)
 
 def fft_impl(x, fft_type, fft_lengths):
   return xla.apply_primitive(fft_p, x, fft_type=fft_type, fft_lengths=fft_lengths)
 
+_complex_dtype = lambda dtype: (onp.zeros((), dtype) + onp.zeros((), onp.complex64)).dtype
+_real_dtype = lambda dtype: onp.zeros((), dtype).real.dtype
+_is_even = lambda x: x % 2 == 0
+
 def fft_abstract_eval(x, fft_type, fft_lengths):
```

Comment:
This is always promoting to the shaped level, whereas if given concrete inputs it could produce a concrete output. It's not incorrect to do it this way, and there's no need to change anything; I just thought I'd mention it because you might be interested if we hadn't talked about this before! (In general an abstract_eval function just needs to return a lattice upper bound for the output given the inputs, and this upper bound isn't as tight as possible.)

Reply:
Interesting, I hadn't really thought about these abstract eval rules before!

In this case, I think manipulating shapes is probably a reasonable thing to do, given that I need to be doing some shape manipulation here.